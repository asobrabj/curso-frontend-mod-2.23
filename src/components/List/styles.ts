import styled from 'styled-components'

interface IListStyle {
  $visible: boolean
}

export const Container = styled.div``
export const Form = styled.form``

export const ListItem = styled.li<IListStyle>`
  padding: 16px;
  font-size: 14px;
  font-weight: 400;
  border-bottom: 1px solid ${(props) => props.theme.color.shadow};
  display: ${(props) => (props.$visible ? 'block' : 'none')};
  transition: opacity 1s ease-in-out;
`

export const List = styled.ul`
  text-transform: capitalize;
  cursor: pointer;
  > .paragraph {
    display: flex;
    justify-content: center;
    font-size: 18px;
    font-weight: bold;
    padding: 16px 0;
    border-bottom: 1px solid ${(props) => props.theme.color.shadow};

    > i {
      margin-right: 5px;
    }
  }
`

export const Button = styled.button`
  border: none;
  font-size: 32px;
  color: ${(props) => props.theme.color.greenDark};
  cursor: pointer;
`
