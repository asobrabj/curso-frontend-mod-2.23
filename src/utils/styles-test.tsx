import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import theme from '../styles/darkTheme'

export const renderTheme = (Children: React.ReactNode) => {
  return render(<ThemeProvider theme={theme}> {Children} </ThemeProvider>)
}
