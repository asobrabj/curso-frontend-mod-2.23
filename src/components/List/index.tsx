import { useState } from 'react'
import { IoIosArrowDown, IoIosArrowForward } from 'react-icons/io'
import { useStorage } from '../../context/localStorage/localStorage'
import * as S from './styles'

interface IListItem {
  name: string
  options: {
    text: string
  }[]
}

export default function List({ name, options }: IListItem) {
  const [isVisible, setIsVisible] = useState<boolean>(false)
  const { getTags } = useStorage()

  const translate = {
    categoria: 'category',
    prioridade: 'priority',
  }

  return (
    <S.Container>
      <S.List>
        <S.ListItem
          className="paragraph"
          $visible
          onClick={() => setIsVisible((v) => !v)}
        >
          <i>{!isVisible ? <IoIosArrowForward /> : <IoIosArrowDown />}</i>
          {name}
        </S.ListItem>

        {options?.length > 0 &&
          options.map((option, index) => (
            <S.ListItem
              $visible={isVisible}
              key={index}
              onClick={() => getTags(translate[name], option.text)}
            >
              {option.text}
            </S.ListItem>
          ))}
      </S.List>
    </S.Container>
  )
}
