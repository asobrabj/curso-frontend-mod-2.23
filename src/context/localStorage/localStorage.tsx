import { createContext, useContext, useEffect, useState } from 'react'
import { IFormData } from '../../components/FormAddAnnotation'
import IList from '../../types/lists'

export interface ILocalStorageContext {
  anotations: Array<IFormData> | null
  setAnotations: (data: IFormData | Array<IFormData>) => void
  getFilters: (value: string) => void
  getTags: (key: 'category' | 'priority', value: string) => void
  getStorage: (key: string) => Array<IFormData | IList> | string | null
  setStorage: (key: string, data: Array<IFormData | IList> | string) => void
}

interface ILocalStorageProvider {
  children: React.ReactNode
}

const LocalStorageContext = createContext<ILocalStorageContext>(
  {} as ILocalStorageContext
)

export function LocalStorageProvider({ children }: ILocalStorageProvider) {
  const [anotations, setAnotations] = useState<Array<IFormData> | null>([])

  useEffect(() => {
    const anotations = getStorage('anotations')
    setAnotations(anotations)
  }, [])

  const getTags = (key: 'category' | 'priority', value: string) => {
    const anotations = getStorage('anotations')
    const data = anotations?.filter(
      (item: IFormData) =>
        item[key]?.toLowerCase().includes(value.toLowerCase())
    )

    if (data) setAnotations(data)
    return
  }

  const getFilters = (value: string) => {
    const anotations = getStorage('anotations')
    const name = anotations?.filter((item: IFormData) =>
      item.name.toLowerCase().includes(value.toLowerCase())
    )

    if (name) setAnotations(name)
    return
  }

  const setAnotationsFn = (data: IFormData | Array<IFormData>) => {
    const storage = getStorage('anotations')
    let newStorage: Array<IFormData> = []
    if (storage) {
      newStorage = [...storage, data]
    }

    if (data instanceof Array) {
      newStorage = data
    }

    setAnotations(newStorage)
    localStorage.setItem('anotations', JSON.stringify(newStorage))
  }

  const getStorage = (key: string) => {
    const storedData = localStorage.getItem(key)
    if (storedData) {
      return JSON.parse(storedData)
    }
    return []
  }

  const saveToStorage = (
    key: string,
    data: Array<IFormData | IList> | string
  ) => {
    localStorage.setItem(key, JSON.stringify(data))
    return
  }
  return (
    <LocalStorageContext.Provider
      value={{
        getTags,
        getFilters,
        anotations,
        setAnotations: setAnotationsFn,
        getStorage,
        setStorage: saveToStorage,
      }}
    >
      {children}
    </LocalStorageContext.Provider>
  )
}

export function useStorage() {
  return useContext(LocalStorageContext)
}
