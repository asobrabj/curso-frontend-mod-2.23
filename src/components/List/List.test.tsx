import { fireEvent, screen } from '@testing-library/react'
import List from '.'
import { useStorage } from '../../context/localStorage/localStorage'
import { renderTheme } from '../../utils/styles-test'

// Mockando o módulo useStorage
jest.mock('../../context/localStorage/localStorage', () => ({
  ...jest.requireActual('../../context/localStorage/localStorage'),
  useStorage: jest.fn(),
}))

describe('List Component', () => {
  it('deve renderizar corretamente', () => {
    const getTagsMock = jest.fn()
    ;(useStorage as jest.Mock).mockReturnValue({ getTags: getTagsMock })

    const testData = {
      name: 'categoria',
      options: [{ text: 'Opção 1' }, { text: 'Opção 2' }],
    }

    const { getByText } = renderTheme(<List {...testData} />)

    expect(getByText(testData.name)).toBeInTheDocument()

    testData.options.forEach((option) => {
      expect(getByText(option.text)).not.toBeVisible()
    })

    fireEvent.click(getByText(testData.name))

    testData.options.forEach((option) => {
      expect(getByText(option.text)).toBeVisible()
    })

    fireEvent.click(getByText(testData.options[0].text))
    expect(getTagsMock).toHaveBeenCalledWith(
      'category',
      testData.options[0].text
    )
  })
  it('deve lidar com propriedades não fornecidas', () => {
    renderTheme(<List name={''} options={[]} />)
    expect(screen.queryByText('categoria')).not.toBeInTheDocument()
  })
  it('snapshot', () => {
    const getTagsMock = jest.fn()
    ;(useStorage as jest.Mock).mockReturnValue({ getTags: getTagsMock })

    const testData = {
      name: 'categoria',
      options: [{ text: 'Opção 1' }, { text: 'Opção 2' }],
    }

    const { container } = renderTheme(<List {...testData} />)
    expect(container).toMatchSnapshot()
  })
})
