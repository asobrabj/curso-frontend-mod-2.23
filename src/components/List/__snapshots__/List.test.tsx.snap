// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Component snapshot 1`] = `
.c1 {
  padding: 16px;
  font-size: 14px;
  font-weight: 400;
  border-bottom: 1px solid #999;
  display: block;
  transition: opacity 1s ease-in-out;
}

.c2 {
  padding: 16px;
  font-size: 14px;
  font-weight: 400;
  border-bottom: 1px solid #999;
  display: none;
  transition: opacity 1s ease-in-out;
}

.c0 {
  text-transform: capitalize;
  cursor: pointer;
}

.c0 >.paragraph {
  display: flex;
  justify-content: center;
  font-size: 18px;
  font-weight: bold;
  padding: 16px 0;
  border-bottom: 1px solid #999;
}

.c0 >.paragraph >i {
  margin-right: 5px;
}

<div>
   
  <div
    class=""
  >
    <ul
      class="c0"
    >
      <li
        class="c1 paragraph"
      >
        <i>
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M294.1 256L167 129c-9.4-9.4-9.4-24.6 0-33.9s24.6-9.3 34 0L345 239c9.1 9.1 9.3 23.7.7 33.1L201.1 417c-4.7 4.7-10.9 7-17 7s-12.3-2.3-17-7c-9.4-9.4-9.4-24.6 0-33.9l127-127.1z"
            />
          </svg>
        </i>
        categoria
      </li>
      <li
        class="c2"
      >
        Opção 1
      </li>
      <li
        class="c2"
      >
        Opção 2
      </li>
    </ul>
  </div>
   
</div>
`;
